tools/dev-upload.sh

```diff
@@ -0,0 +1,74 @@
+#!/bin/bash
+
+set -e
+
+if [ -z "${RELEASE}" ]; then
+  echo "ERROR: RELEASE environment variable needs to be set to the" >&2
+  echo "  target Ubuntu version." >&2
+  exit 1
+fi
+
+MIR_VERSION=$( dpkg-parsechangelog --show-field Version \
+               | perl -n -e '/(^[^~-]+)/ && print $1' )
+GIT_REVISION=$( git rev-parse --short HEAD )
+
+source <( python <<EOF
+import os
+from launchpadlib.launchpad import Launchpad
+
+lp = Launchpad.login_anonymously("mir-ci",
+                                 "production",
+                                 "/tmp/lplib",
+                                 version="devel")
+
+ubuntu = lp.distributions["ubuntu"]
+series = ubuntu.getSeries(name_or_version=os.environ['RELEASE'])
+
+print("UBUNTU_SERIES={}".format(series.name))
+print("UBUNTU_VERSION={}".format(series.version))
+
+dev_ppa = lp.people["mir-team"].getPPAByName(name="dev")
+try:
+  mir = dev_ppa.getPublishedSources(source_name="mir",
+                                    distro_series=series,
+                                    order_by_date=True)[0]
+except IndexError:
+  print("PPA_VERSION=")
+else:
+  print("PPA_VERSION={}".format(mir.source_package_version))
+EOF
+)
+
+if [ -z "${UBUNTU_SERIES}" -o -z "${UBUNTU_VERSION}" ]; then
+  echo "ERROR: \"${RELEASE}\" was not recognized as a valid Ubuntu series" >&2
+  exit 2
+fi
+
+while dpkg --compare-versions "${PPA_VERSION}" ge "${DEV_VERSION}"; do
+  DEV_COUNTER=$(( ${DEV_COUNTER} + 1 ))
+  DEV_VERSION=${MIR_VERSION}~dev-${DEV_COUNTER}
+  if [ ${DEV_COUNTER} -gt 1000 ]; then
+    echo "ERROR: version still lower than PPA after 1000 iterations:" >&2
+    echo "  ${PPA_VERSION} >= ${MIR_VERSION}~-${DEV_COUNTER}" >&2
+    exit 3
+  fi
+done
+
```

Comment:
This whole bit seems unnecessary? We can deterministically generate guaranteed-increasing version numbers from the git tree, right?

The `git describe` output is $LAST_TAG-$COMMITS_SINCE_LAST_TAG-$SHORTHASH. Since our mainline is strictly merge commits, $COMMITS_SINCE_LAST_TAG will be strictly monotone increasing.

Since we're not going to tag before we commit the changelog, I think we could just construct the version as:
`$MIR_VERSION~dev+$COMMITS_SINCE_LAST_TAG`?

Reply:
That does sound plausible, if we decide that git tags are our "version source", we could even just do `git describe` and be done with it.

Any idea how to easily get $COMMITS_SINCE_LAST_TAG, or do you mean to parse `git describe` output?

Reply:
I can't find an obvious way to get git to spit out just $COMMITS_SINCE_LAST_TAG; I was thinking we'd just parse the `git describe` output.

Reply:
The one reason I can think of why VCS isn't nice as the authoritative source on versions is tarballs, which lose that information (other than in the filename).

Reply:
The problem with parsing `git describe` is it will be the tag when we're on that commit... And if our debian/changelog doesn't change when we tag a commit, $COMMITS_SINCE_LAST_TAG will reset and we might have a conflicting version after all.

Reply:
OOOH!

`git rev-list --count HEAD` gets the total number of commits, which is going to be monotone increasing.

Since our policy is to only have merge commits on the mainline, `git rev-list --count --merges HEAD` would do the same thing, but increase more slowly 😸

Reply:
Ack, I went for the former to start with.